react-ui/src/pages/analyzeVotes/PearsonsCorrelation.jsx

```diff
@@ -0,0 +1,133 @@
+    if (!username) { return []; }
+
+    const numberOfEntries = Object.keys(entryPositionLookup).length;
+    const allVotesByUser = votes.reduce((acc, vote) => {
+      if (!acc[vote.username]) {
+        acc[vote.username] = new Array(numberOfEntries);
+      }
+      acc[vote.username][entryPositionLookup[vote.entryId]] = vote.rating;
+      return acc;
+    }, {});
+
+    const userVotes = allVotesByUser[username];
+    // USER DID NOT VOTE
+    if (!userVotes) { return []; }
+
+    return Object.keys(allVotesByUser).filter((u) => u !== username).map((u) => {
+      const otherUser = allVotesByUser[u];
+      let sum1 = 0; let sum2 = 0; let sum1Sq = 0; let sum2Sq = 0; let sumProduct = 0;
+      let n = 0;
+      let numVotes = 0;
+      otherUser.forEach((vote, index) => {
+        const userVote = userVotes[index];
+        if (vote > -1 && userVote > -1) {
+          sum1 += vote;
+          sum2 += userVote;
+          sum1Sq += vote * vote;
+          sum2Sq += userVote * userVote;
+          sumProduct += vote * userVote;
+          n += 1;
+        }
+        if (vote > -1) { numVotes += 1; }
+      });
+      const num = sumProduct - ((sum1 * sum2) / n);
+      const den = Math.sqrt((sum1Sq - ((sum1 * sum1) / n)) * (sum2Sq - ((sum2 * sum2) / n)));
+
+      if (n === 0 || den === 0) { return { username: u, correlation: undefined }; }
+
+      return { username: u, correlation: num / den, group: numVotes < voteMinimum ? GROUP.few : undefined };
+    }).filter((p) => p.correlation !== undefined)
+      .sort((a, b) => a.correlation - b.correlation);
+  }, [username, votes, entryPositionLookup, voteMinimum]);
+
+  const userPoints = pearsons.map((p, i) => ({
+    x: i,
+    y: p.correlation,
+    id: p.username,
+    group: p.username === username2 ? GROUP.selected : (p.group || GROUP.none),
+    text: `User 2: ${trimUsername(p.username)}<br />Pearsons: ${roundTwoDecimals(p.correlation)}`,
+  }));
+
+  const data = createTraces(userPoints, [
+    { name: trimUsername(username2, 20), marker: MARKERS.general.selected },
+    { name: 'Other users', marker: MARKERS.general.unselected },
+    { name: `<${voteMinimum} votes`, marker: MARKERS.general.few },
+  ]);
+
+  const layout = {
+    title: `How similar is ${username} to others?`,
+    xaxis: { title: 'User' },
+    yaxis: { title: 'Similarity (Pearsons)' },
+  };
+
+  const usernames = useMemo(() => pearsons.map((p) => p.username), [pearsons]);
+
+  const handleKeyUp = useCallback((event) => {
+    const { key } = event;
+    if (key === 'ArrowUp' || key === 'ArrowDown') {
+      event.preventDefault();
+      setUsername2((prev) => {
+        const index = usernames.indexOf(prev);
+        if (key === 'ArrowUp') {
+          return usernames[index + 1] || usernames[0];
+        }
+        return usernames[index - 1] || usernames[usernames.length - 1];
+      });
+    }
+  }, [setUsername2, usernames]);
+
+  useEffect(() => {
+    window.addEventListener('keydown', handleKeyUp);
```

Comment:
You should probably use `keyup` in both places.

Reply:
`keydown` helps prevent scroll and gives immediate feedback. I changed the function to `handleKeyDown` instead.